[KOGITO-7278][KSW-Guides] Building Workflow Images with Tekton Pipelines (#333)
apache/incubator-kie-kogito-docs: serverlessworkflow/modules/ROOT/pages/cloud/operator/build-workflow-images-with-tekton.adoc

= Building Workflows Images with Tekton
:compat-mode!:
// Metadata:

> :description: Building Workflows Images with Tekton

Author: ok

:keywords: kogito, workflow, serverless, operator, kubernetes, minikube, openshift, containers, tekton, pipelines
// links

> :kogito_serverless_operator_url: https://github.com/kiegroup/kogito-serverless-operator/

Reviewer: don't we have this env already set on another place?

Author: Remove, was copied from another page but are useless here

> :kogito_greeting_example_url: https://github.com/kiegroup/kogito-examples/tree/stable/serverless-workflow-examples/serverless-workflow-greeting-quarkus

Reviewer: same here.

Author: removed

:kaniko_issue_url: https://github.com/GoogleContainerTools/kaniko/issues/2201

> This document describes how to build a workflow image on a Kubernetes cluster using the Tekton pipeline.

Author: minikube, kind and ocp, added a link to install tekton on plain kubernetes

> This document describes how to build a workflow image on a Kubernetes cluster using the Tekton pipeline.

Author: ok

.Prerequisites

> You hav a git repo containing your workflow and the dockerfile to build the workflow as a docker/podman image.

Reviewer 2: ```suggestion
You have a git repo containing your workflow and the dockerfile to build the workflow as a docker/podman image.
```

Author: ok

[NOTE]
====

> If you are using Openshift you can install the RedHat Pipeline Operator, if you are using minikube you can follow this link:https://github.com/tektoncd/pipeline/blob/main/docs/developers/local-setup.md#using-minikube[tekton guide for Minikube]

Reviewer 2: ```suggestion
If you are using Openshift you can install the Red Hat Pipeline Operator, if you are using minikube you can follow this link:https://github.com/tektoncd/pipeline/blob/main/docs/developers/local-setup.md#using-minikube[tekton guide for Minikube]
```

Author: ok

> If you are using Openshift you can install the RedHat Pipeline Operator, if you are using minikube you can follow this link:https://github.com/tektoncd/pipeline/blob/main/docs/developers/local-setup.md#using-minikube[tekton guide for Minikube]

Reviewer: ```suggestion
If you are using OpenShift you can install the Red Hat Pipeline Operator, if you are using Minikube you can follow this link:https://github.com/tektoncd/pipeline/blob/main/docs/developers/local-setup.md#using-minikube[tekton guide for Minikube]
```

Author: ok

> with Kind you can follow thislink:https://github.com/tektoncd/plumbing/tree/main/hack[Tekton guidem for Kind]

Reviewer 2: ```suggestion
with Kind you can follow this link:https://github.com/tektoncd/plumbing/tree/main/hack[Tekton guide for Kind]
```

Author: ok

> with Kind you can follow thislink:https://github.com/tektoncd/plumbing/tree/main/hack[Tekton guidem for Kind]

Reviewer: ```suggestion
If using Kind you can follow thislink:https://github.com/tektoncd/plumbing/tree/main/hack[Tekton guidem for Kind]
```

Author: ok

====

== Create The pipeline
In the namespace kogito-serverless-operator-system where is deployed the Kogito Serverless Operator in the pipelines menu open the item Pipeline and go inthe create, copy the following yaml.

> It contains a pipelin to clone your image workflow repo, build the image anbd push to the local image registry.

Reviewer 2: ```suggestion
It contains a pipeline to clone your image workflow repo, build the image and push it to the local image registry.
```

Author: ok

in the Pipeline menu, in the item PipelineRun copy the following yaml replacing the placeholder with your values

----
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  generateName: custom-workflow-image-pipeline-run
  namespace: kogito-serverless-operator-system
spec:
  workspaces:
    - name: shared-workspace
      volumeClaimTemplate:
        spec:
          accessModes:
            - ReadWriteOnce
          resources:
            requests:
              storage: 500Mi
  params:
    - name: deployment-name
      value: <your-image-name>
    - name: git-url
      value: https://<your repo>.git
    - name: git-revision
      value: "main"
    - name: IMAGE
      value: "image-registry.openshift-image-registry.svc:5000/kogito-serverless-operator-system/<your-image-name>:latest"
  pipelineRef:
    name: custom-workflow-image-pipeline
----
You can check the execution of the pipeline in the UI,
at the end your image will be available at:
image-registry.openshift-image-registry.svc:5000/kogito-serverless-operator-system/<your-image-name>:latest

== Create The PipelineRun with Tkn cli

> If you have installed the tekton cli you could run the pipeline with this command

Reviewer: ```suggestion
If you have installed the Tekton cli you can run the pipeline with the following command:
```

Author: ok

----
tkn pipeline start kogito-serverless-operator-pipeline \
-w name=shared-workspace,volumeClaimTemplateFile=https://raw.githubusercontent.com/kiegroup/kogito-serverless-operator/main/tekton/volume/persistent_volume.yaml \
-p deployment-name=<your-image-name> \
-p git-url=https://<your repo>.git \
-p git-revision=main \
-p IMAGE='image-registry.openshift-image-registry.svc:5000/kogito-serverless-operator-system/<your-image-name>:latest' \
--use-param-defaults
----

> This command return an id to check the execution with the cli in this way

Reviewer: ```suggestion
This command returns an id to check the execution with the cli in this way
```

Author: ok

----
tkn pipelinerun logs custom-workflow-image-pipeline-run-<id> -f -n <your-namespace>
----

> at the end your image will be available at:

Reviewer: ```suggestion
At the end your image will be cluster's internal registry, like this example:
```

Author: ok

> image-registry.openshift-image-registry.svc:5000/kogito-serverless-operator-system/<your-image-name>:latest

Reviewer: ```suggestion
----
image-registry.openshift-image-registry.svc:5000/kogito-serverless-operator-system/<your-image-name>:latest
----
```

Reviewer: image-registry.openshift-image-registry.svc:5000/ can vary from ocp to ocp cluster.

Author: ok, I'll add the explanation as example
